speculos.py

```diff
@@ -49,11 +49,45 @@ def get_elf_infos(app_path):
             print('[-] failed to find _estack/END_STACK symbol', file=sys.stderr)
             sys.exit(1)
         estack = sym_estack[0]['st_value']
+        supp_ram = elf.get_section_by_name('.rfbss')
+        ram_addr, ram_size = (supp_ram['sh_addr'], supp_ram['sh_size']) if supp_ram is not None else (0, 0)
     stack_size = estack - stack
-    return hex(sh_offset), hex(sh_size), hex(stack), hex(stack_size)
+    return sh_offset, sh_size, stack, stack_size, ram_addr, ram_size
+
+def sanitize_ram_args(model: str, rampage: str, pagesize: str, extra_ram: dict):
+    ''' Check RAM pages probed from main app & libs are identical and throw an error if it's
+    not the case. This is not generic but is sufficient as far as FW Apps team is concerned.
+    (Note: Generic code would merge overlapping pages and remove duplicate ones so that the 
+    list of pages passed to launcher contains unique items).  
+    Also discard extra RAM pages probed from main app or libraries when not running on Blue
+    '''
+    args = []
+    origin = ['*', 'probed']
+    if model == 'blue':
+        if rampage is not None and pagesize is not None:    # then user-provided values supercede probed values
+            (rampage, pagesize) = (int(rampage, 16), int(pagesize, 16))     # Convert to int
+            if extra_ram is None:
+                extra_ram = [{}]
+            # First extra_ram list entry is main app extra RAM params, replace it by user-provided {rampage, pagesize}
+            _ = extra_ram.pop(0)
+            extra_ram.insert(0, {'addr':rampage, 'size':pagesize})
+            origin=['+', 'user-provided']
+
+        if extra_ram is not None:
+            if  len(extra_ram) > 1:     # then all pages shall be identical
+                # Remove identical extra ram params,the python way. Throw an error if resulting list length is not 1
+                extra_ram = [dict(s) for s in set(tuple(d.items()) for d in extra_ram)]
+                if len(extra_ram) > 1:
+                    raise ValueError("Different extra RAM pages for main app and/or libraries!")
+            page = extra_ram[0]
+            print(f'[{origin[0]}] using {origin[1]} {page["size"]:d}-byte additional RAM page @0x{page["addr"]:08x}')
```

Comment:
No need to use string formatting because `hex` returns a string

Reply:
Removed `:d` formatting for page size. Replaced `:08x` formatting with call to hex() for page address.

Reply:
It would actually simplify only the following line in `sanitize_ram_args()` function:
```                extra_ram = [dict(s) for s in set(tuple(d.items()) for d in extra_ram)]```
would be changed to:
```                extra_ram = [tuple(s) for s in set(d for d in extra_ram)]```
or something close to that. But in the rest of the code we'd have to address data in extra_ram through indexes. I don't think this change is worth the loss of readability.

Reply:
Same for the next line please :) (`f'{hex(page["addr"])}'` and `f'{hex(page["size"])}'`)

Reply:
done